Routing rules (#371)
MewsSystems/gitbook-connector-api: operations/routingrules.md

# Routing rules

## Get all routing rules

> Returns all routing rules.

Reviewer: It is correct, because it is one of the required parameters. You either provide a CompanyIds and get all routing rules associated to those, or You provide the specific RoutingRuleIds to get information about those right away.

        {
            "Id": "ff785b22-5422-4d1d-87f4-af2e00b3dfda",
            "CompanyId": "cd441e1a-6f19-4960-887a-af2a00d5d5f8",
            "CompanyRelation": "PartnerCompany",
            "ServiceId": "0907a1b4-ef7a-4aa8-b8a1-af2a00d5ca22",
            "Applicability": "PrepaidOnly",
            "RouteType": "AllStayItems"
        },
        {
            "Id": "d98c9611-0006-4691-a835-af2e00b170c4",
            "CompanyId": "ddc23f8d-131d-44d6-b150-af2a00d5d5f8",
            "CompanyRelation": "TravelAgency",
            "ServiceId": "0907a1b4-ef7a-4aa8-b8a1-af2a00d5ca22",
            "Applicability": "Always",
            "RouteType": "SelectedStayItems",
            "SelectedStayItems":
            {
                "Nights" : true,
                "CityTax" : true,
                "ProductCategoryIds":[
                    "004fa262-7b08-4853-b544-af2a00d5cf1f",
                    "8f0dc6ef-1fd4-44e4-b353-af2c00b24caf",
                    "d4c1b435-0ed1-4d98-9833-af2c00b28518"
                ]
            }
        }
    ],
    "Cursor": "d98c9611-0006-4691-a835-af2e00b170c4"
}
```
| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `RoutingRules` | array of [Routing rules](#routing-rule) | required | Routing rules. |

Reviewer: ```suggestion
| `RoutingRules` | array of [Routing rules](#routing-rule) | required | A collection of Routing rules. |
```

| `Cursor` | string | optional | Unique identifier of the item one newer in time order than the items to be returned. If Cursor is not specified, i.e. null, then the latest or most recent items will be returned. |

#### Routing rule

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Id` | string | required | Unique identifier of the service. |
| `CompanyId` | string | required | Unique identifier of the [Company](companies.md#company) routing rule is assigned to. |
| `CompanyRelation` | string [Company relation](routingrules.md#company-relation) | required | Relation of the company to the routing rule. |

> | `Applicability` | string [Applicability](routingrules.md#applicability) | required | Determines if routing rule apply to all future reservations with this company or travel agency attached or only future reservations that are prepaid in online travel agency (OTA). |

Reviewer: I don't quite like the wording on this one.
```suggestion
| `Applicability` | string [Applicability](routingrules.md#applicability) | required | Determines whether the routing rule applies to all future reservations with this company or travel agency attached or only future reservations that are prepaid in online travel agency (OTA). |
```

> | `Applicability` | string [Applicability](routingrules.md#applicability) | required | Determines if routing rule apply to all future reservations with this company or travel agency attached or only future reservations that are prepaid in online travel agency (OTA). |

Reviewer 2: ```suggestion
| `Applicability` | string [Applicability](routingrules.md#applicability) | required | Determines if routing rule apply to all future reservations with this company or travel agency, or only future reservations that are prepaid. |
```

| `ServiceId` | string | required | Unique identifier of the [Service](services.md#service) routing rule is assigned to. |

> | `RouteType` | string [Route type](routingrules.md#route-type) | required | What should be routed. |

Reviewer: Gotta come up with a better wording, this isn't quite obvious. `What should be routed.` says very little and is too abstract.

> | `SelectedStayItems` | [Selected stay items](routingrules.md#selected-stay-items) | optional | To which stay items routing rule applies to. Returns only if RouteType value is SelectedStayItems. |

Reviewer: Another wording issue, we'll figure it out.

### Company relation

* PartnerCompany
* TravelAgency

### Applicability

* Always
* PrepaidOnly

### Route type

* AllStayItems
* SelectedStayItems

#### Selected stay items

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Nights` | boolean | required | Whether routing rule applies to nights. |
| `CityTax` | boolean | required | Whether routing rule applies to city tax. |

> | `ProductCategoryIds` | array of string | optional | Product categories to which routing rule applies to. |

Reviewer 2: ```suggestion
| `ProductCategoryIds` | array of string | optional | Product categories to which routing rule applies. |
```

### Response

```javascript
{
    "RoutingRules": [
        {
            "Id": "ff785b22-5422-4d1d-87f4-af2e00b3dfda",
            "CompanyId": "cd441e1a-6f19-4960-887a-af2a00d5d5f8",
            "CompanyRelation": "PartnerCompany",
            "ServiceId": "0907a1b4-ef7a-4aa8-b8a1-af2a00d5ca22",
            "Applicability": "PrepaidOnly",
            "RouteType": "AllStayItems"
        },
        {
            "Id": "d98c9611-0006-4691-a835-af2e00b170c4",
            "CompanyId": "ddc23f8d-131d-44d6-b150-af2a00d5d5f8",
            "CompanyRelation": "TravelAgency",
            "ServiceId": "0907a1b4-ef7a-4aa8-b8a1-af2a00d5ca22",
            "Applicability": "Always",
            "RouteType": "SelectedStayItems",
            "SelectedStayItems":
            {
                "Nights" : true,
                "CityTax" : true,
                "ProductCategoryIds":[
                    "004fa262-7b08-4853-b544-af2a00d5cf1f",
                    "8f0dc6ef-1fd4-44e4-b353-af2c00b24caf",
                    "d4c1b435-0ed1-4d98-9833-af2c00b28518"
                ]
            }
        }
    ],

>     "Cursor": "d98c9611-0006-4691-a835-af2e00b170c4"

Reviewer: Cursor shouldn't be here, right?

            "ServiceId": { "Value": "708b3509-69ad-4a92-841b-d81f103edcee" },
            "Applicability": { "Value": "Always" },
            "RouteType": { "Value": "SelectedStayItems" },
            "SelectedStayItems": 
            {
                "Nights" : { "Value":  false },
                "CityTax" : { "Value": true },
                "ProductCategoryIds": { "Value": [
                    "004fa262-7b08-4853-b544-af2a00d5cf1f"
                ]}
            }
        },

    ]
}
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `ClientToken` | string | required | Token identifying the client application. |
| `AccessToken` | string | required | Access token of the client application. |
| `Client` | string | required | Name and version of the client application. |
| `RoutingRules` | array of [Routing rule update parameters](routingrules.md#routing-rule-update-parameters) | required, max 1000 items | Routing rules to be updated. |

#### Routing rule update parameters

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `RoutingRuleId` | string | required | Unique identifier of the routing rule. |

> | `CompanyId` | [String update value](#string-update-value) | optional | Unique identifier of the [Company](companies.md#company) \(or `null` if should not be updated\). |

Reviewer: ```suggestion
| `CompanyId` | [String update value](#string-update-value) | optional | Unique identifier of the [Company](companies.md#company) \(or `null` should it not be updated\). |
```

> | `CompanyRelation` | [String update value](#string-update-value)| optional | [Company relation](routingrules.md#company-relation) to the routing rule \(or `null` if should not be updated\). |

Reviewer: ```suggestion
| `CompanyRelation` | [String update value](#string-update-value)| optional | [Company relation](routingrules.md#company-relation) to the routing rule \(or `null` should it not be updated\). |
```

> | `Applicability` | [String update value](#string-update-value) | required | [Applicability](routingrules.md#applicability) that determines if routing rule apply to all future reservations with this company or travel agency attached or only future reservations that are prepaid in online travel agency (OTA) \(or `null` if should not be updated\). |

Reviewer: Now that I'm thinking, is it correct that `null` should it not be updated? Doesn't providing a null mean nullifying? And not providing not updating? 🤔

| `RouteType` | [String update value](#string-update-value) | required | What should be routed [Route type](routingrules.md#route-type) \(or `null` if should not be updated\). |
| `SelectedStayItems` | [Selected stay items update parameters](routingrules.md#selected-stay-items-update-parameters) | optional | To which stay items routing rule applies to. Returns only if RouteType value is SelectedStayItems. |

#### String update value

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Value` | string | optional | Value which is to be updated. |

#### String array update value

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Value` | array of string | optional | Value which is to be updated. |

#### Bool update value

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Value` | bool | optional | Value which is to be updated. |

#### Selected stay items update parameters

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `Nights` | [Bool update value](#bool-update-value) | optional | Whether routing rule applies to nights \(or `null` if should not be updated\). |

Reviewer: ```suggestion
| `Nights` | [Bool update value](#bool-update-value) | optional | Whether routing rule applies to nights \(or `null` should it not be updated\). |
```

> | `CityTax` | [Bool update value](#bool-update-value) | optional | Whether routing rule applies to city tax \(or `null` if should not be updated\). |

Reviewer: ```suggestion
| `CityTax` | [Bool update value](#bool-update-value) | optional | Whether routing rule applies to city tax \(or `null` should it not be updated\). |
```

> | `ProductCategoryIds` | [String array update value](#string-array-update-value) | optional | Product categories to which routing rule applies to \(or `null` if should not be updated\). |

Reviewer: ```suggestion
| `ProductCategoryIds` | [String array update value](#string-array-update-value) | optional | Product categories to which the routing rule applies to \(or `null` should it not be updated\). |
```

> | `ProductCategoryIds` | [String array update value](#string-array-update-value) | optional | Product categories to which routing rule applies to \(or `null` if should not be updated\). |

Reviewer 2: Small point of grammar - if you use "to which" then you don't need the additional "to" on the end, e.g. "Product categories to which the routing rule applies."

Reviewer: Thanks for this, I'll utilize it. I stopped commenting because I was just about to redo this 😄

### Response

```javascript
{
    "RoutingRules": [
        {
            "Id": "ff785b22-5422-4d1d-87f4-af2e00b3dfda",
            "CompanyId": "35f493d1-6cf7-4132-982a-a5540549c322",
            "CompanyRelation": "PartnerCompany",
            "ServiceId": "708b3509-69ad-4a92-841b-d81f103edcee",
            "Applicability": "Always",
            "RouteType": "AllStayItems"
        },
        {
            "Id": "d98c9611-0006-4691-a835-af2e00b170c4",
            "CompanyId": "73a1fc8e-11c5-4019-b296-9234b651fc68",
            "CompanyRelation": "TravelAgency",
            "ServiceId": "708b3509-69ad-4a92-841b-d81f103edcee",
            "Applicability": "Always",
            "RouteType": "SelectedStayItems",
            "SelectedStayItems":
            {
                "Nights" : false,
                "CityTax" : true,
                "ProductCategoryIds":[
                    "004fa262-7b08-4853-b544-af2a00d5cf1f"
                ]
            }
        }
    ],

>     "Cursor": null

Reviewer: Maybe we should delete this cursor. Even though the request shows it, it is not relevant in a context of update.
